chainstate/src/detail/mod.rs

```diff
@@ -281,101 +287,89 @@ impl<S: BlockchainStorage, V: TransactionVerificationStrategy> Chainstate<S, V>
-    /// is passed to `func` and then returned to the caller.
-    /// Note that the state is reset to `initial_state` on each commit attempt and the returned
-    /// state is the one from the last commit attempt.
-    fn with_rw_tx_and_state<State, MainAction, OnNewAttempt, OnDbErr, Res>(
-        &mut self,
-        initial_state: &State,
-        mut main_action: MainAction,
-        on_new_attempt: OnNewAttempt,
-        on_db_err: OnDbErr,
-    ) -> (State, Result<Res, BlockError>)
-    where
-        State: Clone,
-        MainAction: FnMut(
-            &mut chainstateref::ChainstateRef<TxRw<'_, S>, V>,
-            &mut State,
-        ) -> Result<Res, BlockError>,
-        OnNewAttempt: FnMut(/*attempt_number:*/ usize),
-        OnDbErr: FnOnce(/*attempts_count:*/ usize, chainstate_storage::Error) -> BlockError,
-    {
-        // Note: the purpose of the Option here is just to get rid of extra clone at the beginning.
-        let mut state = None;
-
-        let result = self.with_rw_tx(
-            |chainstate_ref| main_action(chainstate_ref, state.insert(initial_state.clone())),
-            on_new_attempt,
-            on_db_err,
-        );
-
-        // Note: this "or_else" part is only possible in the degenerate case where with_rw_tx
-        // doesn't invoke the closure even once.
-        let state = state.unwrap_or_else(|| initial_state.clone());
-        (state, result)
-    }
-
-    /// Integrate the block into the blocktree, performing all the necessary checks and
-    /// updating `block_status` after each successful check.
+    /// Integrate the block into the blocktree, performing all the necessary checks.
     /// The returned bool indicates whether a reorg has occurred.
     fn integrate_block(
         chainstate_ref: &mut chainstateref::ChainstateRef<TxRw<'_, S>, V>,
         block: &WithId<Block>,
-        block_index: &BlockIndex,
-        block_status: &mut BlockStatus,
-    ) -> Result<bool, BlockError> {
-        chainstate_ref.check_block_parent(block)?;
-        block_status.advance_validation_stage_to(BlockValidationStage::ParentOk);
-
-        chainstate_ref.check_block(block).map_err(BlockError::CheckBlockFailed)?;
-        block_status.advance_validation_stage_to(BlockValidationStage::CheckBlockOk);
-
-        // Note: we have to persist BlockIndex too, because it will be used
-        // by activate_best_chain below. There is no point in saving
-        // an intermediate BlockStatus though, so we ignore block_status here.
-        chainstate_ref.set_new_block_index(block_index)?;
-        chainstate_ref.persist_block(block)?;
+        block_index: BlockIndex,
+    ) -> Result<bool, BlockIntegrationError> {
+        let mut block_status = BlockStatus::new();
+
+        let result = chainstate_ref.check_block(block);
+        if result.is_err() {
+            // FIXME: "technical" errors (e.g. a DB error) should probably not lead
```

Comment:
This is exactly the same problem we talk about with PropertyQueryError. IIUC, you're saying that the same error is returned in different contexts. This is a good example of what happens when an error is just an "error", with no meaning, and then we start throwing it behind try operators `?`; it can easily become a mess where all errors will mean the same thing, which isn't the desired behavior.

So, like I mentioned in the daily, perhaps we should have some kind of recoverable/irrecoverable errors for blocks. But this needs some design work before we commit to it. To elaborate: Generally speaking, some block errors warrant a retry, and others will result in a complete rejection. This distinction isn't possible right now.

Reply:
> IIUC, you're saying that the same error is returned in different contexts.

This is one of the problems but there is a couple more:

1) The meaning of an error is not always clear from its name, so the code that performs an important decision based on it has a chance to misinterpret it. E.g. by looking as `CheckBlockError`'s variants it's hard to tell which of them should lead to permanent block invalidation and which should not. Also I'm afraid that since their names are sometimes generic enough, they may change their meaning slightly in the future, which may change the way they should be interpreted, but this fact will be missed.

2) Those "technical" errors have non-unique representations. For example, I have a  `CheckBlockError` and need to check if it's storage-related. Then first I'll have to look into `CheckBlockError::StorageError`, and `CheckBlockError::PropertyQueryError`. But there is also `CheckBlockError::StateUpdateFailed` that contains `ConnectTransactionError`, which also can be a `StorageError`. Also, `ConnectTransactionError` has a bunch of variants that represent storage-related invariant violations or other broken expectations, like `ConnectTransactionError::BlockIndexCouldNotBeLoaded`. Then, `ConnectTransactionError` also has 
`TransactionVerifierError(TransactionVerifierStorageError)` and `PoSAccountingError(pos_accounting::Error)` variants that can also contain `StorageError`. And I expect that this rabbit hole goes even deeper.
One relatively easy way to deal with it right now would be to implement something similar to the existing `ban_score`. But I'm not sure whether it will be reliable enough either, because unlike ban_score, we're dealing with a permanent decision here.
Another way is to represent the "bad block" result of check_block and its relatives via Ok part of the returned Result, so that the Err part only indicates that the check could not be completed successfully. But this approach will probably complicate error handling, so I'm not sure. Also, it can't be done on all levels, so it's a half-measure anyway.

For now I guess I'll just turn this into a TODO and add an issue. Would it be Ok?

Reply:
I've added a common issue for 3 problems - this one, the one below (about deleting blocks later) and the one about stopping adding transactions back to mempool - https://github.com/mintlayer/mintlayer-core/issues/1033